src/workerd/jsg/async-context.h

```diff
@@ -0,0 +1,166 @@
+  //
+  // Every async resource maintains a reference to the Async Context Frame that was current
+  // at the moment the resource is created.
+  //
+  // Frames form a stack. The default frame is the Root. We "enter" a frame by pushing it
+  // onto to top of the stack (making it "current"), then perform some action within that
+  // frame, then "exit" by popping it back off the stack. The Root is associated with the
+  // Isolate itself such that every isolate always has at least one frame on the stack at
+  // all times. In Node.js terms, the "Async Context Frame" would be most closely aligned
+  // with the concept of an "execution context" or "execution scope".
+  //
+  // Every Frame has a storage context. The current frame determines the currently active
+  // storage context. So, for instance, when we start executing, the Root Frame's storage
+  // context is active. When a timeout elapses and a timer is going to fire, we enter the
+  // timer's Frame which makes that frame's storage context active. Once the timer
+  // callback has completed, we return back to the Root frame and storage context.
+  //
+  // All frames (except for the Root) are created within the scope of a parent, which by
+  // default is whichever frame is current when the new frame is created. When the new frame
+  // is created, it inherits a copy storage context of the parent.
+public:
+  class StorageKey: public kj::Refcounted {
+    // An opaque key that identifies an async-local storage cell within the frame.
+  public:
+    StorageKey() : hash(kj::hashCode(this)) {}
+
+    void reset() { dead = true; }
+    // The owner of the key should reset it when it goes away.
+
+    bool isDead() const { return dead; }
+    inline uint hashCode() const { return hash; }
+    inline bool operator==(const StorageKey& other) const {
+      return hash == other.hash;
+    }
+
+  private:
+    uint hash;
+    bool dead = false;
+  };
+
+  struct StorageEntry {
+    kj::Own<StorageKey> key;
+    Value value;
+
+    inline StorageEntry clone(Lock& js) {
+      return {
+        .key = addRef(*key),
+        .value = value.addRef(js)
+      };
+    }
+  };
+
+  AsyncContextFrame(IsolateBase& isolate);
+  AsyncContextFrame(
+      Lock& js,
+      kj::Maybe<AsyncContextFrame&> maybeParent = nullptr,
+      kj::Maybe<StorageEntry> maybeStorageEntry = nullptr);
+
+  static AsyncContextFrame& current(Lock& js);
+  // Returns the reference to the AsyncContextFrame currently at the top of the stack.
+
+  static kj::Own<AsyncContextFrame> create(
+      Lock& js,
+      kj::Maybe<AsyncContextFrame&> maybeParent = nullptr,
+      kj::Maybe<StorageEntry> maybeStorageEntry = nullptr);
+  // Create a new AsyncContextFrame. If maybeParent is not specified, uses the current().
+  // If maybeStorageEntry is non-null, the associated storage cell in the new frame is
+  // set to the given value.
+
+  static v8::Local<v8::Function> wrap(Lock& js, v8::Local<v8::Function> fn,
```

Comment:
I feel like the name "wrap" is not that intuitive... you're not really wrapping the promise, you're attaching a context to it. Maybe "attach" is a better verb to use?

Reply:
I'm not sure `attach` is any better tho and runs the risk of being confused with kj's `attach` semantics. Maybe `attachContext`?

Reply:
That works for me.